### PR TITLE
Migrate `setup.cfg` to `pyproject.toml`

### PR DESCRIPTION
**Pull request by aaraney**
- Update github actions imports `actions/setup-python` -> `@5`; `actions/checkout` -> `@v4`
- Test all subpackages in ci w/ `3.11` and `3.12`.
- Migrate all subpackage metadata files from `setup.cfg` to `pyproject.toml`.
- Bump all subpackage patch versions

### CODE REVIEW

**Comment by aaraney** [comment]
@jarq6c, [`hydrotools.caches` tests](https://github.com/NOAA-OWP/hydrotools/actions/runs/11921332173/job/33225219879?pr=263) are failing for `python` `3.9`. It looks like a clean up condition that might need a guard clause. I verified that the later versions of python (`>3.9`) in the test matrix are using a later version, `3.10.1`, of `tables`. It looks like `tables` [3.10.0](https://github.com/PyTables/PyTables/releases/tag/v3.10.0) dropped support for `python<3.10`.

**Comment by aaraney** [comment]
@jarq6c, I think the `hydrotools.nwm_client_new` `3.9` tests are failing for a similar reason. My best guess is the wheel pulled for `tables==3.9.2` (version in failing tests) was compiled with `numpy<2` but notice the test runner is pulling `numpy>=2`. So, there are ABI issues when `tables` goes to call `numpy` functions.

If that is the case, we should be able to remedy this by conditionally requiring `numpy<2` for `python<3.10`.

**Comment by aaraney** [comment]
Looks like [9b9fcfa](https://github.com/NOAA-OWP/hydrotools/pull/263/commits/9b9fcfa3c2fad0a41d10c48325568087157bf947) fixed the `hydrotools.nwm_client_new` test failures.

**Comment by aaraney** [comment]
Looks like [17c0090](https://github.com/NOAA-OWP/hydrotools/pull/263/commits/17c00903902003eb45b5d9377bece086a55ac48e) fixed the `hydrotools.caches` test failures.

File: python/nwis_client/pyproject.toml [hunk 1]
```diff
@@ -1,6 +1,46 @@
 [build-system]
+requires = ["setuptools>=70"]
 build-backend = "setuptools.build_meta"
-requires = [
-  "setuptools>=42",
-  "wheel",
+
+[project]
+name = "hydrotools.nwis_client"
+authors = [{ name = "Austin Raney", email = "aaraney@protonmail.com" }]
```
Comment (jarq6c) [inline python/nwis_client/pyproject.toml:7]: It looks like the move to `pyproject.toml` makes it easier to have multiple authors. Would you like to share the authorship? You should remain first author, of course.

**Comment by jarq6c** [comment]
Thanks for tackling all these @aaraney ! I had originally planned to bump the patch versions of all these packages with these changes. What do you think?

**Comment by aaraney** [comment]
@jarq6c, yeah that is cool with me. Ill do that now.

File: python/metrics/src/hydrotools/metrics/_version.py [hunk 2]
```diff
@@ -1 +1 @@
-__version__ = "1.3.5"
+__version__ = "1.3.6"
```
Comment (jarq6c) [inline python/metrics/src/hydrotools/metrics/_version.py:1]: `metrics` was already bumped under #261

Comment (aaraney) [reply python/metrics/src/hydrotools/metrics/_version.py]: Great catch, @jarq6c. Just reverted.

**Review by jarq6c — APPROVED** [review]

**Comment by aaraney** [comment]
Thanks for the review, @jarq6c :tada:!

**Pull request merged**